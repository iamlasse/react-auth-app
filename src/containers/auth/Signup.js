import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import Signup from '../components/auth/Signup';
import { authSel, actions } from '../../store/modules/auth';

const { getAuthFetchStatus, getAuthenticated } = authSel;

const mapStateToProps = state => ({
	fetching: getAuthFetchStatus(state),
	authenticated: getAuthenticated(state)
});

const mapDispatchToProps = dispatch => ({
	...bindActionCreators(
		{
			signupUser: actions.signupRequest,
			loginWith: actions.loginRequestProvider
		},
		dispatch
	)
});

export default connect(mapStateToProps, mapDispatchToProps)(Signup);
