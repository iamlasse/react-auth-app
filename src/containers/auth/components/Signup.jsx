import React from "react";
import PropTypes from "prop-types";
import { Intent } from "@blueprintjs/core";

import { Link } from "react-router-dom";
import { Field, reduxForm } from "redux-form";
import { Heading } from "grommet";
import "@blueprintjs/core/lib/css/blueprint.css";
import "@blueprintjs/icons/lib/css/blueprint-icons.css";
import {
  StyledFormWrapper,
  StyledFormActions,
  StyledButton,
  StyledFormGroup,
  StyledInputGroup,
  StyledLoadingIcon
} from "./Styled";
import AuthButtons from "./AuthButtons";
import Auth from "../../../constants/Auth";

const auth = new Auth();

const validateForm = values => {
  const errors = {};

  errors.email = !values.email
    ? "Email is Requred"
    : !values.email.match(/@/)
    ? "Must be a valid email"
    : undefined;

  errors.password = !values.password
    ? "Email is Requred"
    : !values.password.match(/[A-Z]/)
    ? "Must contain an uppercase letter"
    : values.password.length <= 5
    ? "Must be longer than 5 characters"
    : undefined;

  errors.passwordConfirmation = !values.passwordConfirmation
    ? "Missing password confirmation"
    : values.password !== values.passwordConfirmation
    ? "Passwords must match"
    : undefined;

  return errors;
};

const InputField = ({ input, ...rest }) => (
  <StyledInputGroup {...input} {...rest} />
);

const Signup = ({
  fetching,
  valid,
  pristine,
  dirty,
  handleSubmit,
  loginWith,
  signupUser
}) => {
  const handleSignupnWith = async provider => {
    try {
      const { access_token: accessToken } = await auth.loginWith(provider);
      loginWith(provider, accessToken);
    } catch (error) {
      console.log(error);
    }
  };

  const handleSignupWithEmailPassword = ({ email, password, username }) => {
    signupUser({ username, password, email });
  };

  const intent = Intent.NONE;

  return (
    <StyledFormWrapper>
      <form
        onSubmit={handleSubmit(handleSignupWithEmailPassword)}
        style={{ maxWidth: 300, margin: "auto" }}
      >
        <Heading level={2} style={{ textAlign: "center" }}>
          Create account
        </Heading>
        <AuthButtons handleAction={handleSignupnWith} />
        <StyledFormGroup intent={intent}>
          <Field
            round
            leftIcon="person"
            large
            id="username"
            placeholder="Username"
            intent={intent}
            name="username"
            component={InputField}
          />
          <Field
            name="email"
            round
            leftIcon="envelope"
            large
            id="email"
            placeholder="email@email.com"
            intent={intent}
            component={InputField}
          />
          <Field
            name="password"
            round
            leftIcon="lock"
            large
            id="password"
            type="password"
            placeholder="Password"
            intent={intent}
            component={InputField}
          />

          <Field
            name="passwordConfirmation"
            round
            large
            leftIcon="lock"
            type="password"
            id="passwordConfirmation"
            placeholder="Password Again"
            intent={intent}
            component={InputField}
          />

          <StyledButton
            plain
            primary
            type="submit"
            color="brand-2"
            label=""
            large
            fill
            disabled={!valid && (!dirty || !pristine)}
            loading={fetching}
            onClick={handleSubmit(handleSignupWithEmailPassword)}
          >
            {fetching && (
              <StyledLoadingIcon
                color="white"
                size="small"
                style={{ marginRight: 10 }}
              />
            )}
            Create account with email
          </StyledButton>

          <StyledFormActions>
            Already have an account?
            <Link to="/signin"> Sign In</Link>
          </StyledFormActions>
        </StyledFormGroup>
      </form>
    </StyledFormWrapper>
  );
};

const signupWithReduxForm = reduxForm({
  form: "signup",
  validate: validateForm
})(Signup);

export default signupWithReduxForm;
