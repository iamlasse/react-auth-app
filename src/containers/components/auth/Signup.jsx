import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Button, FormGroup, InputGroup, Intent } from '@blueprintjs/core';
import styled from 'styled-components';
import { Link } from 'react-router-dom';
import { Button as GButton } from 'grommet';
import { GooglePlus, Facebook } from 'grommet-icons';
import '@blueprintjs/core/lib/css/blueprint.css';
import '@blueprintjs/icons/lib/css/blueprint-icons.css';

import Auth from '../../../constants/Auth';
const auth = new Auth();

const StyledFormGroup = styled(FormGroup)`
display: flex;
justify-content: space-around;
`;

const StyledInputGroup = styled(InputGroup)`
margin-bottom: 1em;
`;

const StyledButton = styled(GButton)`
  padding: 8px;
  border-radius: 50px !important;
  color: #fff
	outline: none
	&:active, &:visited {
    outline: none;
  }
`;

class Register extends Component {
	static propTypes = {
		fetching: PropTypes.bool,
		signupUser: PropTypes.func
	};

	static defaultProps = {
		fetching: false,
		signupUser: null
	};

	state = {
		disabled: false,
		// helperText: true,
		inline: false,
		intent: Intent.NONE,
		// label: true,
		// requiredLabel: false,
		username: '',
		password: '',
		passwordConfirmation: ''
	};

	handleSignupnWith = async provider => {
		const { loginWith } = this.props;
		try {
			const { access_token } = await auth.loginWith(provider);
			loginWith(provider, access_token);
		} catch (error) {
			console.log(error);
		}
	};

	handleRegister = e => {
		console.log(e);
		const { username, password, passwordConfirmation } = this.state;
		const { signupUser } = this.props;

		if (password !== passwordConfirmation) return console.error("Passwords don't match");

		if (!username) return console.error('Username empty');

		signupUser(username, password);
		return false;
	};

	handleChange = ({ target: { id, value } }) => {
		this.setState({ [id]: value });
	};

	render() {
		const {
			disabled,
			// helperText,
			inline,
			intent,
			// requiredLabel,
			// label,
			username,
			password,
			passwordConfirmation
		} = this.state;
		const { fetching } = this.props;

		return (
			<div
				style={{
					display: 'flex',
					justifyContent: 'center',
					alignItems: 'center',
					height: '100vh',
					width: '100vw',
					backgroundColor: '#efefef'
				}}
			>
				<form onSubmit={this.handleRegister} style={{ maxWidth: 300, margin: 'auto' }}>
					<h2 style={{ textAlign: 'center' }}>Create account</h2>
					<div className="auth-buttons">
						<StyledButton
							plain
							primary
							focusIndicator={false}
							label="Sign up with Google"
							color={'#ea4335'}
							icon={<GooglePlus color="white" />}
							margin="xsmall"
							fill
							loading={fetching}
							onClick={() => this.handleSignupnWith('google')}
						/>

						<StyledButton
							plain
							primary
							focusIndicator={false}
							label="Sign up with Facebook"
							color={'#3b5998'}
							icon={<Facebook color="white" />}
							margin="xsmall"
							fill
							loading={fetching}
							onClick={() => this.handleSignupnWith('facebook')}
						/>
					</div>
					<StyledFormGroup disabled={disabled} inline={inline} intent={intent}>
						<StyledInputGroup
							round
							leftIcon="person"
							large
							id="username"
							placeholder="Username"
							disabled={disabled}
							intent={intent}
							onChange={this.handleChange}
						/>
						<StyledInputGroup
							round
							large
							leftIcon="lock"
							type="password"
							id="password"
							placeholder="Password"
							disabled={disabled}
							intent={intent}
							onChange={this.handleChange}
						/>
						<StyledInputGroup
							round
							large
							leftIcon="lock"
							type="password"
							id="passwordConfirmation"
							placeholder="Password Again"
							disabled={disabled}
							intent={intent}
							onChange={this.handleChange}
						/>
						<div>
							<Button
								style={{ borderRadius: 50 }}
								disabled={!(username && password && passwordConfirmation)}
								intent="primary"
								large
								fill
								loading={fetching}
								onClick={this.handleRegister}
							>
								Register
							</Button>
						</div>
						<span
							style={{
								display: 'flex',
								alignItems: 'center',
								marginTop: 20,
								justifyContent: 'space-between'
							}}
						>
							Already have an account?
							<Link to="/signin"> Sign In</Link>
						</span>
					</StyledFormGroup>
				</form>
			</div>
		);
	}
}

export default Register;
