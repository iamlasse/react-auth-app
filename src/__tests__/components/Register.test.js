import React from 'react';
import { shallow } from 'enzyme';
import Signup from '../../containers/auth/components/Signup';

let wrapper;
describe('Register Component', () => {
	beforeEach(() => {
		wrapper = shallow(<Signup />);
	});

	it('renders without crashing', () => {
		expect(wrapper).toMatchSnapshot();
	});
});
