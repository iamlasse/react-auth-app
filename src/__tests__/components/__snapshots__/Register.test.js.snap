// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register Component renders without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ForwardRef
    destroyOnUnmount={true}
    enableReinitialize={false}
    forceUnregisterOnUnmount={false}
    form="signup"
    getFormState={[Function]}
    keepDirtyOnReinitialize={false}
    persistentSubmitErrors={false}
    pure={true}
    shouldAsyncValidate={[Function]}
    shouldError={[Function]}
    shouldValidate={[Function]}
    shouldWarn={[Function]}
    touchOnBlur={true}
    touchOnChange={false}
    updateUnregisteredFields={false}
    validate={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "destroyOnUnmount": true,
      "enableReinitialize": false,
      "forceUnregisterOnUnmount": false,
      "form": "signup",
      "forwardedRef": null,
      "getFormState": [Function],
      "keepDirtyOnReinitialize": false,
      "persistentSubmitErrors": false,
      "pure": true,
      "shouldAsyncValidate": [Function],
      "shouldError": [Function],
      "shouldValidate": [Function],
      "shouldWarn": [Function],
      "touchOnBlur": true,
      "touchOnChange": false,
      "updateUnregisteredFields": false,
      "validate": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "destroyOnUnmount": true,
        "enableReinitialize": false,
        "forceUnregisterOnUnmount": false,
        "form": "signup",
        "forwardedRef": null,
        "getFormState": [Function],
        "keepDirtyOnReinitialize": false,
        "persistentSubmitErrors": false,
        "pure": true,
        "shouldAsyncValidate": [Function],
        "shouldError": [Function],
        "shouldValidate": [Function],
        "shouldWarn": [Function],
        "touchOnBlur": true,
        "touchOnChange": false,
        "updateUnregisteredFields": false,
        "validate": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
